cpu/sam0_common/sam0-eth/eth-netdev.c

```diff
@@ -0,0 +1,172 @@
+#include "net/ethernet.h"
+#include "net/netdev/eth.h"
+
+#include "periph/gpio.h"
+
+#include "sam0_eth_netdev.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+#include "log.h"
+
+/* Internal helpers */
+int sam0_eth_init(void);
+int sam0_eth_send(const struct iolist *iolist);
+int sam0_eth_receive_blocking(char *data, unsigned max_len);
+void sam0_eth_set_mac(const char *mac);
+void sam0_eth_get_mac(char *out);
+
+sam0_eth_netdev_t _sam0_eth_dev;
+
+static int _sam0_eth_init(netdev_t *netdev)
+{
+    (void)netdev;
+    sam0_eth_init();
+    return 0;
+}
+
+static void _sam0_eth_isr(netdev_t *netdev)
+{
+    netdev->event_callback(netdev, NETDEV_EVENT_RX_COMPLETE);
+    return;
+}
+
+static int _sam0_eth_recv(netdev_t *netdev, void *buf, size_t len, void *info)
+{
+    (void)info;
+    (void)netdev;
+    mutex_lock(&_sam0_eth_dev.dev_lock);
+    unsigned ret = sam0_eth_receive_blocking((char *)buf, len);
+    mutex_unlock(&_sam0_eth_dev.dev_lock);
+
+    return ret;
+}
+
+static int _sam0_eth_send(netdev_t *netdev, const iolist_t *iolist)
+{
+    int ret;
+
+    mutex_lock(&_sam0_eth_dev.dev_lock);
+
+    netdev->event_callback(netdev, NETDEV_EVENT_TX_STARTED);
+    ret = sam0_eth_send(iolist);
+    if (ret == -EOVERFLOW) {
+        mutex_unlock(&_sam0_eth_dev.dev_lock);
+        /* TODO: use a specific netdev callback here ? */
+        return -EOVERFLOW;
+    }
+    netdev->event_callback(netdev, NETDEV_EVENT_TX_COMPLETE);
+    mutex_unlock(&_sam0_eth_dev.dev_lock);
+    return ret;
+}
+
+static int _sam0_eth_get(netdev_t *netdev, netopt_t opt, void *val, size_t max_len)
+{
+    int res = -1;
+
+    switch (opt) {
+        case NETOPT_ADDRESS:
+            assert(max_len >= ETHERNET_ADDR_LEN);
+            sam0_eth_get_mac((char *)val);
+            res = ETHERNET_ADDR_LEN;
+            break;
+        default:
+            res = netdev_eth_get(netdev, opt, val, max_len);
+            break;
+    }
+
+    return res;
+}
+
+static int _sam0_eth_set(netdev_t *netdev, netopt_t opt, const void *val, size_t max_len)
+{
+    int res = -1;
+
+    switch (opt) {
+        case NETOPT_ADDRESS:
+            assert(max_len >= ETHERNET_ADDR_LEN);
+            sam0_eth_set_mac((char *)val);
+            res = ETHERNET_ADDR_LEN;
+            break;
+        default:
+            res = netdev_eth_set(netdev, opt, val, max_len);
+            break;
+    }
+
+    return res;
+}
+
+static const netdev_driver_t _sam0_eth_driver =
+{
+    .send = _sam0_eth_send,
+    .recv = _sam0_eth_recv,
+    .init = _sam0_eth_init,
+    .isr  = _sam0_eth_isr,
+    .get  = _sam0_eth_get,
+    .set  = _sam0_eth_set,
+};
+
+void sam0_eth_setup(netdev_t* netdev)
+{
+
+    LOG_INFO("[sam0_eth]: initializing SAM0 Ethernet MAC (GMAC) device\n");
+
+    /* initialize locking */
+    mutex_init(&_sam0_eth_dev.dev_lock);
+
+    _sam0_eth_dev.netdev = netdev;
+    /* set the netdev driver */
+    netdev->driver = &_sam0_eth_driver;
+}
+
+/* TODO: rework the whole isr management... */
```

Comment:
What's wrong with it?

Reply:
Everything, I would say. I am not completely happy with this. I think the error cases can be manage a bit better.